src/optimizer/query_rewrite.c

```diff
@@ -3499,6 +3500,651 @@ qo_reduce_outer_joined_tables (PARSER_CONTEXT * parser, PT_NODE * spec, PT_NODE
   return query;
 }
 
+/*
+ * qo_reduce_joined_referenced_tables () - Removes the table of the primary key from the join
+ *                                         with the table with the foreign key referencing it.
+ *   return: void
+ *   parser(in): parser context
+ *   query(in): query to check
+ *
+ * Note: The table with the primary key can be removed from the join with the table with the foreign key referencing it
+ *       if there is no other reference than the joining purpose. This is because the relationship
+ *       between the primary key and the foreign key replaces the data filter of the join. If the join of the table
+ *       with the primary key is removed, predicates for IS NOT NULL must be added because it cannot be filtered
+ *       for NULL.
+ *
+ *    e.g. drop if exists child, parent;
+ *         create table parent (c1 int primary key, c2 int);
+ *         create table child (c1 int, parent_c1 int references parent (c1), c2 int);
+ *
+ *         select c.* from child c inner join parent p on p.c1 = c.parent_c1 where c.c2 = 1;
+ *         select c.* from child c, parent p where c.parent_c1 = p.c1 and c.c2 = 1;
+ *           -> select c.* from child c where c.c2 = 1 and c.parent_c1 is not null;
+ *
+ *         select c.* from child c inner join parent p on p.c1 = c.parent_c1 where p.c2 = 1;
+ *         select c.* from child c, parent p where c.parent_c1 = p.c1 and p.c2 = 1;
+ *           -> do not change.
+ */
+static void
+qo_reduce_joined_referenced_tables (PARSER_CONTEXT * parser, PT_NODE * query)
+{
+  PT_NODE *curr_pk_spec = NULL, *prev_pk_spec = NULL, *next_pk_spec = NULL;
+  PT_NODE *curr_fk_spec = NULL;
+  SPEC_CNT_INFO info = { NULL, 0, 0, NULL };
+  PT_NODE *backup_where = NULL;
+  PT_NODE *curr_pred = NULL, *prev_pred = NULL, *next_pred = NULL, *parent_pred = NULL;
+  PT_NODE *exclude_spec_point_list = NULL, *exclude_spec_point = NULL, *exclude_spec = NULL;
+  PT_NODE *pred_point_list = NULL, *pred_point = NULL;
+  PT_NODE *curr_pred_point = NULL, *prev_pred_point = NULL, *next_pred_point = NULL;
+  PT_NODE *append_pred_list = NULL;
+  PT_NODE *curr_append_pred = NULL, *prev_append_pred = NULL, *next_append_pred = NULL;
+  int i;
+
+  if (parser == NULL || query == NULL || query->node_type != PT_SELECT)
+    {
+      return;
+    }
+
+  if (query->info.query.q.select.hint & PT_HINT_NO_MERGE)
+    {
+      return;
+    }
+
+  if (query->info.query.q.select.where != NULL && query->info.query.q.select.where->or_next != NULL)
+    {
```

Comment:
it can not check below
```
where col1 = 1 and (col2 = 1 or col3 = 1)
```

Reply:
Below, if it is not PT_EQ, it is skipped.
Since PT_OR is not PT_EQ, it is skipped.

I will add it to the test case.

Reply:
PLZ add or_next check

Reply:
Added it.